design/application-abi.md

```diff
@@ -15,26 +15,42 @@ Current Unstable ABI
 There are two kinds of modules:
 
  - A *command* exports a function named `_start`, with no arguments and no return
-   values. Environments shall call this function once, after instantiating the
-   module and all of its dependencies. After this function exits, the instance
-   is considered terminated and no further use of any of its exports should be
-   permitted.
+   values.
 
- - A *reactor* exports a function named `_initialize`, with no arguments and no
-   return values. Environments shall call this function once, after instantiating
-   the module and all of its dependencies. After this function exits, the instance
-   remains live, and its exports may be accessed.
+   Command instances may assume that they will be called from the environment
+   at most once. Command instances may assume that none of their exports are
+   accessed outside the duraction of that call.
+
+ - All other modules are *reactors*. A reactor may export a function named
+   `_initialize`, with no arguments and no return values.
```

Comment:
Why not *must* export?    Isn't more explicit to say you either need `_start` or `_initialize` otherwise you are not wasi module at all?   This might make it easier to distinguish wasi modules in the wild?  

This existing verbiage also makes it sounds like requirement.

Reply:
I didn't see a need to distinguish between a module with an `_initialize` which does nothing, and a module with no `_initialize`. Am I missing something?

Reply:
I guess I was thinking, as a wasi runtime I'd like to be able to reject non-wasi modules if neither `_start` not `_initialize` are found:  "This doesn't look like a wasi module.. I don't know how to start it".

e.g: https://github.com/WebAssembly/wabt/blob/master/src/interp/interp-wasi.cc#L725

It also allows the linker to use `--entry=_initialize` and report an error if crt1. o is missing.   

But I feel like both of those reason are pretty weak.   Your counterargument would be size cost of the exported empty function?

Reply:
Maybe if you think of it more like: "This doesn't look like a wasi command or reactor.. I don't know how to start it".

Reply:
Right, but if we don't require `_initialize` to exist then any wasm module with a function export (i.e. all wasm modules) will look like reactors, so that error cannot practically be reported.

Reply:
I’m going to hijack the conversation just a little and pose this question instead: will we permit modules which have neither of those entry points but export some functionality which internally uses some of the WASI syscalls such as `fd_read` to be correctly hooked up as WASI module in the embedder and those syscalls correctly resolved? I’m thinking here mainly of situations where a module doesn’t necessarily have to use the preopens and doesn’t rely on syscalls that require access to any. I guess you could even think of such a module as a lib instead of a WASI “binary”.

Reply:
Yes we have been using the term "reactor" to refer to this library-style use case.

The question at hand is .. if  a reactor doesn't need to do any initialization (e.g. pre-opens, static constructors, libc init) should it still be expected an empty `_initialize` function so the runtime and knows it is a wasi reactor.

I would imagine that the number of useful reactors that don't need to do *any* initialization are fairly small.     Certainly any reactor that uses libc at all needs to call the libc init functions.

Reply:
I expect we'll see lots of reactors that don't do any initialization :-). libc's environment variables are now lazy-initialized, and the only other libc init is the preopen code, which not all reactors will link in.

I don't yet see the advantage of distinguishing between a "wasi module" and a "non-wasi module". If a module doesn't have an `_initialize` function then it's just a plain wasm module that we can instantiate, if we can satisfy its imports.

Reply:
OK.. I don't feel strong about this so lgtm.

Note that any program with C++ static constructor will also have `_initialize` ..its not just libc itself.  I think any program that uses cout or err ends up in this category.   But I agree for C and maybe rust programs that might be less common.